modules/dnn/src/layers/region_layer.cpp

```diff
@@ -317,6 +317,7 @@ class RegionLayerImpl CV_FINAL : public RegionLayer
                                 dstData[box_index + 1] = (y + y_tmp) / rows;
                                 dstData[box_index + 2] = (srcData[box_index + 2]) * (srcData[box_index + 2]) * 4 * biasData[2 * a] / wNorm;
                                 dstData[box_index + 3] = (srcData[box_index + 3]) * (srcData[box_index + 3]) * 4 * biasData[2 * a + 1] / hNorm;
+                                dstData[box_index + 4] = srcData[p_index];
 
                                 scale = srcData[p_index];
                                 if (classfix == -1 && scale < thresh)
```

Comment:
If I understand correctly, `new_coords` only have affects on how we calculate the width and height of bboxes. It does not relate to whether objectness should be kept or not. See these lines from darknet: https://github.com/AlexeyAB/darknet/blob/62e5549cef53bd11890808615e0bfae59cfd4491/src/yolo_layer.c#L132-L153

By they way, the `scale` in line 322 should be the objectness as you mentioned. Objectness is only used to calculate the final object score, and hence we do not need to keep it in the output.

Reply:
1. The the `dstData[index + 4]` is uninitialized if `new_coords=1`, but initialized if `new_coords=0`. I think it is not good.
2. If you have `dstData[index + 4]` values, you can optimize NMS by check if `dstData[index + 4] > conf_thres`, no need to find the maximum of all classes probabilities.